### ISSUE TITLE
Rest for pattern matching positional record fields

### ISSUE THREAD

**Issue by water-mizuu** [dart-lang/language#3190]
Inspired by #3189

I just thought about indeed how inconvenient it was for collection items to be destructured irrefutably, and while I like the explicitness that a rest List/Map pattern supports, I thought, why not allow it for positional record fields?

```dart
var (int x,int  y, ...) = (4, 2, 1, 0); // (4, 2)
```

If it is possible, then we may even get a record type for the rest.

```dart
var (int x, int y, ...(int, int, int) rest) = (8, 4, 2, 1, 0);
```

Now for the questions. Would named record fields also work here? What would be the problems of this syntax?

**Comment by lrhn** [comment]
It's not impossible.

The `(int x, int y, ...)` does make it easier to ignore parts of a known record type, but it would only be _allowed_ to work on known record type. We need to know the fields that are ignored.

So `switch (o as dynamic) { case (int x, int y, ...): ...}` would just not be allowed. If the "matched value type" is not a record type, you can't omit record fields from the match, because then the check case needs to first do an `is` check to see if it *is* a record with a compatible type, and there is no functionality in the language to do that without knowing a specific record shape.

When allowed, it is just be a shorthand for filling in `_, _` or named `name: _` for each known record field that doesn't matter.


Capturing individual values *into* a record is more tricky.
Again, not impossible.

We do something similar for lists, `case [var first, ... var middle, var last]` which creates a new list from the middle elements. It's the _one_ place where a pattern creates a *new* object itself (but obviously calling getters in object patters can create new objects too, this is just the only place where the pattern does it for you).

Doing something similar for records could work. It would still require a known record shape, otherwise it's just not allowed. 
With a statically known record type, `(var v1, var v2, var :n1, ... var rest)` would take all the fields of the known record type not otherwise matched by the pattern, and create a new record with the remaining positional fields (in original shape order) and remaining named fields. The shape and type of that record would be known at compile-time.

(You can get the same effect today by introducing a getter, say:
```dart
typedef $int5 = (int, int, int, int, int);
extension on $int5 {
  (int, int, int) get $last3 => ($3, $4, $5);
}
void main () {
  var ((int x, int y, _, _, _) && $int5($last3: rest)) = (8, 4, 2, 1, 0);
  // or
  //  var $int5($1: x, $2: y, $last3: rest) = (8, 4, 2, 1, 0);
  print(rest); // (2, 1, 0)
}
```
That suggests a desugaring, and proves that the feature wouldn't provide any real power, just a more convenient syntax for something you can already do.

**Comment by water-mizuu** [comment]
I agree about it being restricted to known record types, since it would need to be inferred, not to mention the unknown amount of fields. However, could they then just be inferred as supertype `Record`? What about if the pattern had explicit types? Or can the language not do type checking at runtime? So for example,

```dart
switch(o as dynamic) {
    case (int x, int y, ...Record rest): // Of course, this would mean that types would be lost at compile time, but dynamic *is* used.
    case (int x, int y, ...(int, int, int) rest): // What if we made the type explicit? Would that still not work?
}
```

**Comment by lrhn** [comment]
The `...(int, int, int) rest` could probably work. It states unambiguously what the shape of "the rest" must be, so we can deduce what the shape of "everything" must be.

The `...Record rest` cannot work. You can think of each record shape as a separate class, a class which is only part of the program if the record shape occurs. In order to match `(int x, int y, ...Record rest)`, it must first check if the runtime type of `o` is a record with ... some unknown shape that starts with two `int`s. There is no support for that. You can only check whether it has a specific, known shape. Here the shape to check for is unknown, so no check can be made.
(And we do _not_ want to check against *all* possible shapes that start with two positional fields. Because even if we did that for every such record shape which occurs anywhere in the program, which itself requires a whole-world compilation strategy, we would also have to be able to create an instance of the "rest" of each of those record types.)
